activation/poet.go

```diff
@@ -272,3 +268,112 @@ func (c *HTTPPoetClient) req(ctx context.Context, method, path string, reqBody,
+type PoetClient struct {
+	db             poetDbAPI
+	id             []byte
+	logger         *zap.Logger
+	client         HTTPPoetClient
+	requestTimeout time.Duration
+
+	// Used to avoid concurrent requests for proof.
+	gettingProof sync.Mutex
+	// cached members of the last queried proof
+	proofMembers map[string][]types.Hash32
+}
+
+func newPoetClient(db poetDbAPI, server types.PoetServer, cfg PoetConfig, logger *zap.Logger) (*PoetClient, error) {
+	client, err := NewHTTPPoetClient(server, cfg, WithLogger(logger))
+	if err != nil {
+		return nil, fmt.Errorf("creating HTTP poet client: %w", err)
+	}
+	poetClient := &PoetClient{
+		db:             db,
+		id:             server.Pubkey.Bytes(),
+		logger:         logger,
+		client:         *client,
+		requestTimeout: cfg.RequestTimeout,
+		proofMembers:   make(map[string][]types.Hash32, 1),
+	}
+
+	return poetClient, nil
+}
+
+func (c *PoetClient) Address() string {
+	return c.client.Address()
+}
+
+func (c *PoetClient) Submit(
+	ctx context.Context,
+	deadline time.Time,
+	prefix, challenge []byte,
+	signature types.EdSignature,
+	nodeID types.NodeID,
+) (*types.PoetRound, error) {
+	logger := c.logger.With(
+		log.ZContext(ctx),
+		zap.String("poet", c.Address()),
+		log.ZShortStringer("smesherID", nodeID),
+	)
+
+	logger.Debug("querying for poet pow parameters")
+	powCtx, cancel := withConditionalTimeout(ctx, c.requestTimeout)
+	defer cancel()
+	powParams, err := c.client.PowParams(powCtx)
+	if err != nil {
+		return nil, &PoetSvcUnstableError{msg: "failed to get PoW params", source: err}
+	}
+
+	logger.Debug("doing pow with params", zap.Any("pow_params", powParams))
+	startTime := time.Now()
+	nonce, err := shared.FindSubmitPowNonce(
+		ctx,
+		powParams.Challenge,
+		challenge,
+		nodeID.Bytes(),
+		powParams.Difficulty,
+	)
+	metrics.PoetPowDuration.Set(float64(time.Since(startTime).Nanoseconds()))
+	if err != nil {
+		return nil, fmt.Errorf("running poet PoW: %w", err)
+	}
+
+	logger.Debug("submitting challenge to poet proving service")
+
+	submitCtx, cancel := withConditionalTimeout(ctx, c.requestTimeout)
+	defer cancel()
+	return c.client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, PoetPoW{
+		Nonce:  nonce,
+		Params: *powParams,
+	})
+}
+
+func (c *PoetClient) Proof(ctx context.Context, roundID string) (*types.PoetProof, []types.Hash32, error) {
+	c.gettingProof.Lock()
+	defer c.gettingProof.Unlock()
+
+	if members, ok := c.proofMembers[roundID]; ok {
+		if proof, err := c.db.GetProofForRound(c.id, roundID); err == nil {
+			c.logger.Debug("returning cached proof", zap.String("round_id", roundID))
+			return proof, members, nil
+		}
+	}
+
+	getProofsCtx, cancel := withConditionalTimeout(ctx, c.requestTimeout)
+	defer cancel()
+	proof, members, err := c.client.Proof(ctx, roundID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting proof: %w", err)
+	}
+
+	if err := c.db.ValidateAndStore(getProofsCtx, proof); err != nil && !errors.Is(err, ErrObjectExists) {
```

Comment:
The timeout should only be for the request to poet, not for storing the result (compare with previous code).

```suggestion
	if err := c.db.ValidateAndStore(ctx, proof); err != nil && !errors.Is(err, ErrObjectExists) {
```